File: AOC_2022/Day 14/14.py
import re

class Solution:
    def __init__(self):
        self.grid = [['.' for _ in range(1000)] for _ in range(200)]
        self.copy_grid = []
        self.wall_index = 0

    def init(self, lines):
        pattern = re.compile(r"(\d+),(\d+)")
        for line in lines:
            matches = re.findall(pattern, line)
            prev = (int(matches[0][0]), int(matches[0][1]))
            self.wall_index = max(self.wall_index, prev[1])

            for match in matches[1:]:
                curr = (int(match[0]), int(match[1]))
                if prev[0] == curr[0]:
                    for i in range(min(prev[1], curr[1]), max(prev[1], curr[1]) + 1):
                        self.grid[i][prev[0]] = '#'
                else:
                    for i in range(min(prev[0], curr[0]), max(prev[0], curr[0]) + 1):
                        self.grid[prev[1]][i] = '#'
                prev = curr
                self.wall_index = max(self.wall_index, prev[1])

        self.wall_index += 2
        self.copy_grid = [row[:] for row in self.grid]

    def dfs(self, pos):
        if pos[0] == len(self.grid) - 1 or self.grid[0][500] == 'o':
            return False

        if self.grid[pos[0] + 1][pos[1]] == '.':
            return self.dfs((pos[0] + 1, pos[1]))

        if self.grid[pos[0] + 1][pos[1]] == '#' or self.grid[pos[0] + 1][pos[1]] == 'o':
            if self.grid[pos[0] + 1][pos[1] - 1] == '.':
                return self.dfs((pos[0], pos[1] - 1))
            elif self.grid[pos[0] + 1][pos[1] + 1] == '.':
                return self.dfs((pos[0], pos[1] + 1))
            else:
                self.grid[pos[0]][pos[1]] = 'o'
                return True

        return False

    def part1(self):
        result = 0
        self.grid[0][500] = '+'
        while self.dfs((0, 500)):
            result += 1
        print(f"Part 1: {result}")

    def part2(self):
        result = 0
        self.grid = [row[:] for row in self.copy_grid]
        self.grid[0][500] = '+'
        self.grid[self.wall_index] = ['#' for _ in range(len(self.grid[0]))]
        while self.dfs((0, 500)):
            result += 1
        print(f"Part 2: {result}")


def main():
    file = open("input.txt", "r")
    lines = file.read().splitlines()
    s = Solution()
    s.init(lines)
    s.part1()
    s.part2()


if __name__ == "__main__":
    main()

File: AOC_2022/Day 14/14.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <regex>

using namespace std;

class Solution
{
    public:
    vector<vector<char>> grid, copyGrid;
    int wallIndex = 0;
    void init(ifstream& input){
        string line;
        grid.resize(200, vector<char>(1000, '.'));
        regex pattern(R"((\d+),(\d+))");
        while (getline(input, line))
        {
            sregex_iterator it(line.begin(), line.end(), pattern), end;
            pair<int, int> prev = {stoi(it->str(1)), stoi(it->str(2))};
            it++;
            wallIndex = max(wallIndex, prev.second);
            while(it != end){
                pair<int, int> curr = {stoi(it->str(1)), stoi(it->str(2))};
                if(prev.first == curr.first){
                    for(int i = min(prev.second, curr.second); i <= max(prev.second, curr.second); i++){
                        grid[i][prev.first] = '#';
                    }
                }
                else{
                    for(int i = min(prev.first, curr.first); i <= max(prev.first, curr.first); i++){
                        grid[prev.second][i] = '#';
                    }
                }
                prev = curr;
                wallIndex = max(wallIndex, prev.second);
                it++;
            }
        }
        wallIndex += 2;
        copyGrid = grid;
    }

    bool dfs(pair<int, int> pos){
        if(pos.first == grid.size() - 1 || grid[0][500] == 'o'){
            return false;
        }
        if(grid[pos.first+1][pos.second] == '.'){
            return dfs({pos.first + 1, pos.second});
        }
        if(grid[pos.first + 1][pos.second] == '#' || grid[pos.first+1][pos.second] == 'o'){
            if(grid[pos.first+1][pos.second - 1] == '.'){
               return dfs({pos.first, pos.second - 1});
            }
            else if(grid[pos.first+1][pos.second + 1] == '.'){
                return dfs({pos.first, pos.second + 1});
            }
            else{
                grid[pos.first][pos.second] = 'o';
                return true;
            }
        }
        return false;
    }

    void part1(){
        int result = 0;
        grid[0][500] = '+';
        while(dfs({0, 500})){
            result++;
        }
        cout<<"Part 1: "<<result<<"\n";
    }
    void part2(){
        int result = 0;
        grid = copyGrid;
        grid[0][500] = '+';
        for(int j = 0; j < grid[0].size(); j++){
            grid[wallIndex][j] = '#';
        }
        while(dfs({0, 500})){
            result++;
        }
        cout<<"Part 2: "<<result<<"\n";
    }

};

int main(){
    ifstream in("input.txt");
    if (!in.is_open())
    {
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}